torch/nested/_internal/nested_tensor.py

```diff
@@ -83,6 +102,45 @@ def __repr__(self):
             grad_fn_str = f", grad_fn={self.grad_fn}"
         return f"NestedTensor(size={self._size}, offsets={self.offsets}{grad_fn_str})"
 
+    def __tensor_flatten__(self):
+        return ["_values", "_offsets"], (self.requires_grad, self.ragged_size)
+
+    @staticmethod
+    def __tensor_unflatten__(inner_tensors: Dict, meta):
+        assert len(inner_tensors) == 2
+        values = inner_tensors["_values"]
+        offsets = inner_tensors["_offsets"]
+        (requires_grad, ragged_size, *extra_meta) = meta
+
+        if len(extra_meta) > 0:
+            (source,) = extra_meta
+            # Avoid circular import
+            from torch._dynamo.source import TensorProperty, TensorPropertySource
+
+            shape_env = offsets.shape[0].node.shape_env
+            sym_ragged_size = shape_env.create_symintnode(
+                shape_env.create_symbol(
+                    ragged_size,
+                    TensorPropertySource(source, TensorProperty.SIZE, 1),
+                ),
+                hint=ragged_size,
+            )
```

Comment:
@bdhirsh can you review this?

But this looks suspicious

Reply:
I think `__tensor_unflatten__()` is not the right place to allocate the extra symint.

In the steady-state, we don't want this to allocate a fresh symint on every round trip of flatten + unflatten:
```
attrs, meta = t.__tensor_flatten__()
new_t = type(t).__tensor_unflatten([getattr(t, x) for x in attrs], meta)
```

So... when we call `__tensor_unflatten__(inner_ts, meta)`, we probably want `meta` to already have the extra needed symint, so `__tensor_unflatten__` can just do the dummy "create an instance of the subclass and dump all of the internal tensors/metadata into it" without having to allocate anything.

Reply:
Two options that seem (hopefully) reasonable to me are:

(1) Just special case `meta_utils.py` to do the extra symint allocation directly and have that be the extra context

(2) Figure out a more general mechanism for meta_utils.py to query the subclass and ask it do to extra work (like constructing an extra symint).

If we're already stuck with hardcoding `is_nested` in the fakeifying logic, then (1) seems ok

Reply:
Makes sense! Updating to go with (1）

Reply:
Just out of curiousity, what are situations where we might have multiple round trips of flatten + unflatten? My current understanding is that I go into meta_utils to fakify once, and then when I do operations, I am responsible for doing unwrap/rewrap in my torch dispatch function.

Reply:
We go through it once to initially fakeify in dynamo, yep.

And we'll go through it again at least once in AOTAutograd: during functionalization, AOTAutograd will replace every input `SubclassTensor(FakeTensor)` with a `SubclassTensor(FunctionalTensor(FakeTensor))`: https://github.com/pytorch/pytorch/pull/104483/files#diff-df954bbf954d2dcb81f687876053267ffa4ddb36ed86b7d2bd76319ff2b94416R469.

We also unflatten at runtime in AOTAutograd, when taking the "plain tensor" outputs of the compiled graph, and wrapping them back up into subclasses to return to the user (if the user's captured graph expects outputs to be tensor subclasses).